irc/metadata.go

```diff
@@ -0,0 +1,257 @@
+package irc
+
+import (
+	"errors"
+	"regexp"
+	"slices"
+	"strings"
+
+	"maps"
+
+	"github.com/ergochat/ergo/irc/caps"
+	"github.com/ergochat/ergo/irc/modes"
+	"github.com/ergochat/ergo/irc/utils"
+)
+
+var (
+	errMetadataTooManySubs = errors.New("too many subscriptions")
+)
+
+type MetadataStore = map[string]string
+
+type MetadataThing = struct {
+	Target     string
+	Key        string
+	Visibility string // currently unused
+	Value      string
+}
+
+func metadataSet(server *Server, session *Session, target string, key string, value string) {
+	targetChannel := server.channels.Get(target)
+	targetClient := server.clients.Get(target)
+
+	if targetChannel != nil {
+		targetChannel.stateMutex.Lock()
+
+		targetChannel.metadata[key] = value
```

Comment:
Make this symmetrical with the client case IMO: don't precreate the map, check if it's `nil` here and create it if necessary